`x-partial` warnings with GHC 9.8

We have 3 warnings about `Data.List`'s `head` and `tail`, but I guess we can nor should do anything other than switching the warning off for these files:
```
src/Data/ListLike/Base.hs:608:12: warning: [GHC-63394] [-Wx-partial]
    In the use of ‘head’
    (imported from Data.List, but defined in GHC.List):
    "This is a partial function, it throws an error on empty lists. Use pattern matching or Data.List.uncons instead. Consider refactoring to use Data.List.NonEmpty."
    |
608 |     head = L.head
    |            ^^^^^^

src/Data/ListLike/Base.hs:610:12: warning: [GHC-63394] [-Wx-partial]
    In the use of ‘tail’
    (imported from Data.List, but defined in GHC.List):
    "This is a partial function, it throws an error on empty lists. Replace it with drop 1, or use pattern matching or Data.List.uncons instead. Consider refactoring to use Data.List.NonEmpty."
    |
610 |     tail = L.tail
    |            ^^^^^^

src/Data/ListLike/DList.hs:53:25: warning: [GHC-63394] [-Wx-partial]
    In the use of ‘tail’
    (imported from Data.List, but defined in GHC.List):
    "This is a partial function, it throws an error on empty lists. Replace it with drop 1, or use pattern matching or Data.List.uncons instead. Consider refactoring to use Data.List.NonEmpty."
   |
53 |   tail = UnsafeDList . (List.tail .) . unsafeApplyDList
   |                         ^^^^^^^^^
```

--- a/.github/workflows/haskell-ci.yml
+++ b/.github/workflows/haskell-ci.yml
@@ -8,9 +8,9 @@
 #
 # For more information, see https://github.com/andreasabel/haskell-ci
 #
-# version: 0.17.20230928
+# version: 0.17.20231010
 #
-# REGENDATA ("0.17.20230928",["github","ListLike.cabal"])
+# REGENDATA ("0.17.20231010",["github","ListLike.cabal"])
 #
 name: Haskell-CI
 on:
@@ -32,11 +32,11 @@ jobs:
     strategy:
       matrix:
         include:
-          - compiler: ghc-9.8.0.20230919
+          - compiler: ghc-9.8.1
             compilerKind: ghc
-            compilerVersion: 9.8.0.20230919
+            compilerVersion: 9.8.1
             setup-method: ghcup
-            allow-failure: true
+            allow-failure: false
           - compiler: ghc-9.6.3
             compilerKind: ghc
             compilerVersion: 9.6.3
@@ -146,7 +146,7 @@ jobs:
           echo "HCNUMVER=$HCNUMVER" >> "$GITHUB_ENV"
           echo "ARG_TESTS=--enable-tests" >> "$GITHUB_ENV"
           echo "ARG_BENCH=--enable-benchmarks" >> "$GITHUB_ENV"
-          if [ $((HCNUMVER >= 90800)) -ne 0 ] ; then echo "HEADHACKAGE=true" >> "$GITHUB_ENV" ; else echo "HEADHACKAGE=false" >> "$GITHUB_ENV" ; fi
+          echo "HEADHACKAGE=false" >> "$GITHUB_ENV"
           echo "ARG_COMPILER=--$HCKIND --with-compiler=$HC" >> "$GITHUB_ENV"
           echo "GHCJSARITH=0" >> "$GITHUB_ENV"
         env:
@@ -175,18 +175,6 @@ jobs:
           repository hackage.haskell.org
             url: http://hackage.haskell.org/
           EOF
-          if $HEADHACKAGE; then
-          cat >> $CABAL_CONFIG <<EOF
-          repository head.hackage.ghc.haskell.org
-             url: https://ghc.gitlab.haskell.org/head.hackage/
-             secure: True
-             root-keys: 7541f32a4ccca4f97aea3b22f5e593ba2c0267546016b992dfadcd2fe944e55d
-                        26021a13b401500c8eb2761ca95c61f2d625bfef951b939a8124ed12ecf07329
-                        f76d08be13e9a61a377a85e2fb63f4c5435d40f8feb3e12eb05905edb8cdea89
-             key-threshold: 3
-          active-repositories: hackage.haskell.org, head.hackage.ghc.haskell.org:override
-          EOF
-          fi
           cat >> $CABAL_CONFIG <<EOF
           program-default-options
             ghc-options: $GHCJOBS +RTS -M3G -RTS
@@ -238,9 +226,6 @@ jobs:
           if [ $((HCNUMVER >= 80200)) -ne 0 ] ; then echo "    ghc-options: -Werror=missing-methods" >> cabal.project ; fi
           cat >> cabal.project <<EOF
           EOF
-          if $HEADHACKAGE; then
-          echo "allow-newer: $($HCPKG list --simple-output | sed -E 's/([a-zA-Z-]+)-[0-9.]+/*:\1,/g')" >> cabal.project
-          fi
           $HCPKG list --simple-output --names-only | perl -ne 'for (split /\s+/) { print "constraints: $_ installed\n" unless /^(ListLike)$/; }' >> cabal.project.local
           cat cabal.project
           cat cabal.project.local

--- a/CHANGELOG.md
+++ b/CHANGELOG.md
@@ -1,6 +1,11 @@
 CHANGES
 =======
 
+#### 4.7.8.2 (2023-10-10)
+
+  - pacify GHC 9.8's new `x-partial` warning
+  - tested with GHC 7.10 - 9.8
+
 #### 4.7.8.1 (2023-07-12)
 
   - allow `bytestring-0.12` and fix its deprecation warnings

--- a/ListLike.cabal
+++ b/ListLike.cabal
@@ -1,7 +1,6 @@
 Cabal-Version: >= 1.10
 Name: ListLike
-Version: 4.7.8.1
-x-revision: 2
+Version: 4.7.8.2
 
 License: BSD3
 Maintainer: David Fox <dsf@seereason.com>, Andreas Abel
@@ -25,7 +24,7 @@ Description: Generalized support for list-like structures in Haskell.
 Stability: Stable
 
 Tested-With:
-  GHC == 9.8.0
+  GHC == 9.8.1
   GHC == 9.6.3
   GHC == 9.4.7
   GHC == 9.2.8

--- a/src/Data/ListLike/Base.hs
+++ b/src/Data/ListLike/Base.hs
@@ -605,9 +605,11 @@ instance ListLike [a] a where
     cons x l = x : l
     snoc l x = l ++ [x]
     append = (++)
-    head = L.head
+    -- Andreas Abel, 2023-10-10, issue #32:
+    -- Use implementation of 'head' and 'tail' in terms of 'uncons' to avoid the x-partial warning under GHC 9.8
+    uncons []       = Nothing
+    uncons (x : xs) = Just (x, xs)
     last = L.last
-    tail = L.tail
     init = L.init
     null = L.null
     length = L.length

--- a/src/Data/ListLike/DList.hs
+++ b/src/Data/ListLike/DList.hs
@@ -50,7 +50,9 @@ instance ListLike (DList a) a where
   -- The following tail function restores the spirit of difference
   -- lists, at the cost of breaking data abstraction, i.e.,
   -- using the constructor and destructor of the newtype DList.
-  tail = UnsafeDList . (List.tail .) . unsafeApplyDList
+  -- Andreas Abel, 2023-10-10, issue #32:
+  -- Use @drop 1@ instead of @tail@ as the latter triggers the x-partial warning in GHC 9.8.
+  tail = UnsafeDList . (List.drop 1 .) . unsafeApplyDList
 #else
   tail = D.tail
 #endif

--- a/testsrc/TestInfrastructure.hs
+++ b/testsrc/TestInfrastructure.hs
@@ -271,10 +271,11 @@ instance LL.FoldableLL (MyList a) a where
     foldl1 f (MyList x) = foldl1 f x
 
 instance Sem.Semigroup (MyList a) where
-  (<>) = mappend
+    MyList x <> MyList y = MyList (x ++ y)
+
 instance Monoid (MyList a) where
     mempty = MyList []
-    mappend (MyList x) (MyList y) = MyList (x ++ y)
+    mappend = (Sem.<>)
 
 instance IsList (MyList a) where
     type Item (MyList a) = a
@@ -283,8 +284,10 @@ instance IsList (MyList a) where
 
 instance LL.ListLike (MyList a) a where
     singleton x = MyList [x]
-    head (MyList x) = head x
-    tail (MyList x) = MyList (tail x)
+    -- Andreas Abel, 2023-10-10, issue #32:
+    -- Avoid 'head' and 'tail' from 'Prelude' and thus the x-partial warning of GHC 9.8.
+    head (MyList (x:xs)) = x
+    tail (MyList (x:xs)) = MyList xs
     null (MyList x) = null x
 
 instance IsString (MyList Char) where

--- a/testsrc/runtests.hs
+++ b/testsrc/runtests.hs
@@ -58,9 +58,13 @@ prop_length f = LL.length f == length (LL.toList f)
 prop_cons f i = llcmp (LL.cons i f) (i : (LL.toList f))
 prop_append f1 f2 = llcmp (LL.append f1 f2) (LL.toList f1 ++ LL.toList f2)
 prop_head f = not (LL.null f) ==> LL.head f == head (LL.toList f)
-prop_last f = not (LL.null f) ==> LL.last f == last (LL.toList f)
+  where head (x:xs) = x
+  -- Andreas Abel, 2023-10-10, issue #32:
+  -- Redefine 'head' and 'tail' to avoid the x-partial warning of GHC 9.8.
 prop_tail f = not (LL.null f) ==> llcmp (LL.tail f) (tail (LL.toList f))
+  where tail (x:xs) = xs
 prop_init f = not (LL.null f) ==> llcmp (LL.init f) (init (LL.toList f))
+prop_last f = not (LL.null f) ==> LL.last f == last (LL.toList f)
 prop_null f = LL.null f == null (LL.toList f)
 prop_length2 f = checkLengths f (LL.toList f)
 prop_length3 f1 f2 = llcmp (LL.append f1 f2) (LL.toList f1 ++ LL.toList f2)

